File: packages/core/src/channels/feishu.ts
/**
 * Feishu (Lark) Channel Client
 *
 * Sends messages via Feishu Open Platform API
 */

import type {
  FeishuChannelConfig,
  SendMessageOptions,
  SendMessageResult,
  TestChannelResult,
} from "./types";

const FEISHU_API_BASE = "https://open.feishu.cn/open-apis";

interface TenantAccessTokenResponse {
  code: number;
  msg: string;
  tenant_access_token?: string;
  expire?: number;
}

/**
 * Get tenant access token from Feishu
 */
async function getTenantAccessToken(
  config: FeishuChannelConfig
): Promise<{ token?: string; error?: string }> {
  try {
    const response = await fetch(
      `${FEISHU_API_BASE}/auth/v3/tenant_access_token/internal`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          app_id: config.app_id,
          app_secret: config.app_secret,
        }),
      }
    );

    const data: TenantAccessTokenResponse = await response.json();

    if (data.code !== 0 || !data.tenant_access_token) {
      return { error: data.msg || "Failed to get access token" };
    }

    return { token: data.tenant_access_token };
  } catch (error) {
    return {
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Determine receive_id_type based on the chat ID format
 */
function getReceiveIdType(chatId: string): string {
  if (chatId.startsWith("ou_")) {
    return "open_id";
  }
  if (chatId.startsWith("on_")) {
    return "union_id";
  }
  if (chatId.startsWith("oc_")) {
    return "chat_id";
  }
  if (chatId.includes("@")) {
    return "email";
  }
  // Default to chat_id
  return "chat_id";
}

/**
 * Send a message to Feishu
 */
export async function sendFeishuMessage(
  config: FeishuChannelConfig,
  options: SendMessageOptions
): Promise<SendMessageResult> {
  if (!config.app_id || !config.app_secret) {
    return { success: false, error: "App ID and App Secret are required" };
  }

  if (!options.chatId) {
    return {
      success: false,
      error: "Chat ID (open_id, chat_id, or email) is required",
    };
  }

  // Get access token first
  const { token, error: tokenError } = await getTenantAccessToken(config);
  if (!token) {
    return { success: false, error: tokenError || "Failed to get access token" };
  }

  try {
    const receiveIdType = getReceiveIdType(options.chatId);

    const response = await fetch(
      `${FEISHU_API_BASE}/im/v1/messages?receive_id_type=${receiveIdType}`,
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          receive_id: options.chatId,
          msg_type: "text",
          content: JSON.stringify({ text: options.message }),
        }),
      }
    );

    const data = await response.json();

    if (data.code !== 0) {
      return {
        success: false,
        error: data.msg || `Code ${data.code}`,
      };
    }

    return {
      success: true,
      messageId: data.data?.message_id,
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Test Feishu channel configuration
 */
export async function testFeishuChannel(
  config: FeishuChannelConfig
): Promise<TestChannelResult> {
  if (!config.app_id || !config.app_secret) {
    return { success: false, error: "App ID and App Secret are required" };
  }

  // Test by getting tenant access token
  const { token, error } = await getTenantAccessToken(config);

  if (!token) {
    return { success: false, error: error || "Failed to get access token" };
  }

  return {
    success: true,
    details: "Successfully obtained tenant access token",
  };
}

File: crates/viben-core/src/channels/feishu.rs
//! Feishu (Lark) Channel Client
//!
//! Sends messages via Feishu Open Platform API

use super::types::{FeishuConfig, SendMessageOptions, SendMessageResult, TestChannelResult};
use reqwest::Client;
use serde::{Deserialize, Serialize};

const FEISHU_API_BASE: &str = "https://open.feishu.cn/open-apis";

#[derive(Debug, Serialize)]
struct TenantTokenRequest {
    app_id: String,
    app_secret: String,
}

#[derive(Debug, Deserialize)]
struct TenantTokenResponse {
    code: i32,
    msg: Option<String>,
    tenant_access_token: Option<String>,
    #[allow(dead_code)]
    expire: Option<i64>,
}

#[derive(Debug, Serialize)]
struct SendMessageRequest {
    receive_id: String,
    msg_type: String,
    content: String,
}

#[derive(Debug, Deserialize)]
struct SendMessageResponse {
    code: i32,
    msg: Option<String>,
    data: Option<SendMessageData>,
}

#[derive(Debug, Deserialize)]
struct SendMessageData {
    message_id: Option<String>,
}

/// Get tenant access token from Feishu
async fn get_tenant_access_token(config: &FeishuConfig) -> Result<String, String> {
    let app_id = match &config.app_id {
        Some(id) if !id.is_empty() => id.clone(),
        _ => return Err("App ID is required".to_string()),
    };

    let app_secret = match &config.app_secret {
        Some(secret) if !secret.is_empty() => secret.clone(),
        _ => return Err("App Secret is required".to_string()),
    };

    let client = Client::new();

    let request_body = TenantTokenRequest {
        app_id,
        app_secret,
    };

    let response = match client
        .post(format!("{}/auth/v3/tenant_access_token/internal", FEISHU_API_BASE))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => return Err(e.to_string()),
    };

    let data: TenantTokenResponse = match response.json().await {
        Ok(d) => d,
        Err(e) => return Err(format!("Failed to parse response: {}", e)),
    };

    if data.code != 0 || data.tenant_access_token.is_none() {
        return Err(data.msg.unwrap_or_else(|| "Failed to get access token".to_string()));
    }

    Ok(data.tenant_access_token.unwrap())
}

/// Determine receive_id_type based on the chat ID format
fn get_receive_id_type(chat_id: &str) -> &'static str {
    if chat_id.starts_with("ou_") {
        "open_id"
    } else if chat_id.starts_with("on_") {
        "union_id"
    } else if chat_id.starts_with("oc_") {
        "chat_id"
    } else if chat_id.contains('@') {
        "email"
    } else {
        // Default to chat_id
        "chat_id"
    }
}

/// Send a message to Feishu
pub async fn send_feishu_message(
    config: &FeishuConfig,
    options: &SendMessageOptions,
) -> SendMessageResult {
    // Get access token first
    let token = match get_tenant_access_token(config).await {
        Ok(t) => t,
        Err(e) => return SendMessageResult::error(e),
    };

    if options.chat_id.is_empty() {
        return SendMessageResult::error("Chat ID (open_id, chat_id, or email) is required");
    }

    let client = Client::new();
    let receive_id_type = get_receive_id_type(&options.chat_id);

    // Build message content (text type)
    let content = serde_json::json!({ "text": options.message }).to_string();

    let request_body = SendMessageRequest {
        receive_id: options.chat_id.clone(),
        msg_type: "text".to_string(),
        content,
    };

    let response = match client
        .post(format!(
            "{}/im/v1/messages?receive_id_type={}",
            FEISHU_API_BASE, receive_id_type
        ))
        .header("Authorization", format!("Bearer {}", token))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => return SendMessageResult::error(e.to_string()),
    };

    let data: SendMessageResponse = match response.json().await {
        Ok(d) => d,
        Err(e) => return SendMessageResult::error(format!("Failed to parse response: {}", e)),
    };

    if data.code != 0 {
        return SendMessageResult::error(
            data.msg.unwrap_or_else(|| format!("Code {}", data.code))
        );
    }

    SendMessageResult::success(
        data.data.and_then(|d| d.message_id)
    )
}

/// Test Feishu channel configuration
pub async fn test_feishu_channel(config: &FeishuConfig) -> TestChannelResult {
    // Test by getting tenant access token
    match get_tenant_access_token(config).await {
        Ok(_) => TestChannelResult::success("Successfully obtained tenant access token"),
        Err(e) => TestChannelResult::error(e),
    }
}
